src/overflow.ts

```diff
@@ -0,0 +1,73 @@
+import { AttributePart, noChange } from 'lit-html';
+import { Directive, directive, DirectiveResult } from 'lit-html/directive.js';
+
+type OnOverflow = (
+	visibleElements: Set<HTMLElement>,
+	overflownElements: Set<HTMLElement>,
+) => void;
+
+function isEntryHidden(el: IntersectionObserverEntry) {
+	return el.boundingClientRect.height === 0;
+}
+
+class OverflowDirective extends Directive {
+	_observer?: IntersectionObserver;
+	_container?: HTMLElement | null;
+
+	render() {
+		return noChange;
+	}
+
+	update(part: AttributePart, [onOverflow]: [OnOverflow]) {
+		const slot = part.element as HTMLSlotElement;
+		this._container = slot.parentElement;
+
+		slot.addEventListener('slotchange', () => {
+			this.overflow(slot, onOverflow);
+		});
+
+		return noChange;
+	}
+
+	overflow(element: HTMLSlotElement, onOverflow: OnOverflow) {
+		this._observer?.disconnect();
+
+		const elements = Array.from(element.assignedElements()) as HTMLElement[];
+		const visibleElements: Set<HTMLElement> = new Set();
+		const overflownElements: Set<HTMLElement> = new Set();
+
+		this._observer = new IntersectionObserver(
+			(entries) => {
+				entries.forEach((entry) => {
+					const el = entry.target as HTMLElement;
+
+					if (entry.intersectionRatio === 1) {
+						visibleElements.add(el);
+						overflownElements.delete(el);
+					} else if (isEntryHidden(entry)) {
+						visibleElements.delete(el);
+						overflownElements.delete(el);
+					} else {
+						visibleElements.delete(el);
+						overflownElements.add(el);
+					}
+
+					onOverflow(visibleElements, overflownElements);
+				});
```

Comment:
You are calling the callback for every entry. It should be sufficient to call it once.

```suggestion
				entries.forEach((entry) => {
					const el = entry.target as HTMLElement;

					if (entry.intersectionRatio === 1) {
						visibleElements.add(el);
						overflownElements.delete(el);
					} else if (isEntryHidden(entry)) {
						visibleElements.delete(el);
						overflownElements.delete(el);
					} else {
						visibleElements.delete(el);
						overflownElements.add(el);
					}

				});
				onOverflow(visibleElements, overflownElements);
```